doc/storage.conf.5
.\" Process this file with
.\" groff -man -Tascii storage.conf.5
.\"
.TH STORAGE.CONF 5 "DECEMBER 2014" RozoFS "User Manuals"
.SH NAME
storage.conf \- RozoFS storage daemon configuration file
.SH DESCRIPTION
.B "storage.conf"
is managed thanks to 
.B libconfig
<http://www.hyperrealm.com/libconfig/> and thus complies to the
.B libconfig
grammar.

.SS listen (mandatory)

Specifies list of IP(s) (or hostname(s)) and port(s) the storio process must listen to for receiving write and read requests from clients.

    addr: (IP or hostname that the server should listens to)
            - must be a valid IPv4 address or a valid hostname.
            - if "*" is specified storio will listen on any of the interfaces.

    port: (TCP port number (in decimal) to use for storio process)
    In single storio mode, only one storio is started that listens on this port to service every configured cluster.
    In multiple storio mode, one storio is started per cluster that listen on this port + the cluster number to service the given cluster only.

.SS self-healing (optional)

Gives the delay in minutes before starting an automatic rebuild of a broken device by relocating the data on the remaining devices. The lowest possible value is 1. When self-healing is configured, the 
.I export-hosts 
parameter must be configured too. 

.SS export-hosts

This parameter is only needed when self healing is configured. It gives the list of addresses (or dns names) where exportd daemon is running. When securing the exportd thanks to virtual IP address only the virtual IP address is declared (i.e "192.168.10.1"). When no virtual IP address is used the list of "physical" IP addresses of each host securing the exportd must be given (i.e "192.168.10.1/192.168.20.1").
 
.SS storages (mandatory)

It's the list of local storage managed by this storaged.
Each storage is identified by the cluster to which belongs the storage (cid) 
and by it's identifier in the cluster (sid). Each pair (cid,sid) must be
unique. The root is the path where data will be stored for this storage.
Under this path stands the logical devices (numbered from 0) on which the
storage will spread the data. The storage writes a new data file on the 
less occupied device. A mapping file stores the chosen device for each
file. The device holding the mapping file is computed from a hash. Although
the number of devices can be increased to scale the storage up, the number
of mapping devices can not change (because of the hash). The mapping files 
have some replica written on the devices following their main device.

    cid: (cluster identifier)
            - must be an integer.
            - should exist in exportd configuration file.

    sid: (storage identifier)
            - must be an integer (between 1 and 255).
            - should exist in exportd configuration file.

    root: (directory path where data will be stored)
            - must be an existing directory.
            - must be unique for each storage managed by this storaged.
	    
    device-total: (total number of devices)
            - must be an integer.
	    
    device-mapper: (number of devices used for file to device mapping)
            - must be an integer.
            - must be lower than or equal to device-total.
            - must never be changed.
	    
    device-redundancy: (number of copies of each file to device mapping file)
            - must be an integer.
            - must be lower than or equal to device-mapper.	    
            - must not be decreased.


.SH EXAMPLE
.PP
.nf
.ta +3i
# sample RozoFS storage configuration file

self-healing = 10;
export-hosts = "192.168.0.1/192.168.0.2";

listen = (
    {addr = "192.168.1.1"; port = 41001; },
    {addr = "192.168.2.1"; port = 41001; }
);

storages = (
    {cid = 1; sid = 1; root = "/srv/rozofs/storages/storage_1_1"; device-total = 6; device-mapper = 3; device-redundancy = 3;},
    {cid = 1; sid = 2; root = "/srv/rozofs/storages/storage_1_2"; device-total = 6; device-mapper = 3; device-redundancy = 3;},
    {cid = 2; sid = 1; root = "/srv/rozofs/storages/storage_2_1"; device-total = 3; device-mapper = 3; device-redundancy = 3;}
 );

.SH FILES
.I /etc/rozofs/storage.conf (/usr/local/etc/rozofs/storage.conf)
.RS
The system wide configuration file.
.\".SH ENVIRONMENT
.\".SH DIAGNOSTICS
.\".SH BUGS
.SH AUTHOR
Fizians <http://www.fizians.com>
.SH "SEE ALSO"
.BR rozofs (7),
.BR storaged (8)


debian/changelog
rozofs (1.6.alpha4) stable; urgency=low

  * (rozofs) See git log for changes

 -- Fizians S.A.S. <devel@rozofs.com>  Mon, 01 Jan 2014 00:00:00 +0100


manager/rozofs/cli/storage.py
# -*- coding: utf-8 -*-
#
# Copyright (c) 2013 Fizians SAS. <http://www.fizians.com>
# This file is part of Rozofs.
#
# Rozofs is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published
# by the Free Software Foundation, version 2.
#
# Rozofs is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see
# <http://www.gnu.org/licenses/>.

import sys
from rozofs.core.platform import Platform, Role
from rozofs.core.storaged import ListenConfig
from rozofs.core.agent import ServiceStatus
from rozofs.cli.output import puts
from rozofs.cli.output import ordered_puts
from collections import OrderedDict

def list(platform, args):
    
    configurations = {}
    
    for h, n in platform._nodes.items():
        if n.has_one_of_roles(Role.STORAGED):
            configurations[h] = n.get_configurations(Role.STORAGED)

    sid_l={}
    for stor in configurations:
        sid_l[stor] = []
        lid_l={}
        if configurations[stor] is None:
            sid_l.update({stor:"not reachable"})
            continue
        for lconfig in configurations[stor][Role.STORAGED].listens:
            lid_l = OrderedDict([
                ('addr', lconfig.addr),
                ('port', lconfig.port)
            ])
            sid_l[stor].append(lid_l)

    ordered_puts(sid_l)

def get(platform, args):

    for host in args.nodes:
        if not host in platform.list_nodes(Role.STORAGED):
                raise Exception('%s: invalid storaged server.' % host)

    sid_l={}
    for host, configuration in platform.get_configurations(args.nodes,
            Role.STORAGED).items():

            sid_l[host]=[]
            lid_l={}
            if configuration is None:
                sid_l.update({host:"not reachable"})
                continue
            for lconfig in configuration[Role.STORAGED].listens:
                lid_l = OrderedDict([
                                     ('addr', lconfig.addr),
                                     ('port', lconfig.port)
                                     ])
                sid_l[host].append(lid_l)

    ordered_puts(sid_l)

def add(platform, args):
    e_host = platform._active_export_host

    for host in args.nodes:
        if not host in platform.list_nodes(Role.STORAGED):
            raise Exception('%s: invalid storaged server.' % host)

    for host in args.nodes:
        
        configurations = platform._get_nodes(e_host)[host].get_configurations(Role.STORAGED)
        configuration = configurations[Role.STORAGED]
        for listener in configuration.listens:
            # if given interface is '*', remove existing interfaces
            if args.interface == "*":
                configuration.listens = []
                continue
            elif args.interface == listener.addr:
                if args.port == listener.port:
                    raise Exception('entry %s:%s already exists.' %
                            (args.interface, args.port))
            if listener.addr == '*':
                configuration.listens = []
        sid_l={}
        sid_l[host]=[]
        lid_l={}
        lconfig = ListenConfig(args.interface, args.port)
        configuration.listens.append(lconfig)
        configurations[Role.STORAGED] = configuration
        platform._get_nodes(e_host)[host].set_configurations(configurations)

        for lconfig in configuration.listens:
            lid_l = OrderedDict([
                ('addr', lconfig.addr),
                ('port', lconfig.port)
            ])
            sid_l[host].append(lid_l)

        ordered_puts(sid_l)

def remove(platform, args):
    e_host = platform._active_export_host

    for host in args.nodes:
        if not host in platform.list_nodes(Role.STORAGED):
            raise Exception('%s: invalid storaged server.' % host)
    
    for host in args.nodes:
        configurations = platform._get_nodes(e_host)[host].get_configurations(Role.STORAGED)
        configuration = configurations[Role.STORAGED]
        check = True

        for listener in configuration.listens:
            if args.interface == listener.addr:
                if args.port == listener.port:
                    # listen entry is found
                    check = False
                    # Check if it's the last listen entry
                    # Replaced by default address
                    if len(configuration.listens) == 1:
                        listener.addr = "*"
                        listener.port = 41001
                    else:
                        configuration.listens.remove(listener)
        if check:
            raise Exception('entry %s:%s does not exist.' % (args.interface,
            args.port))
        sid_l={}
        sid_l[host]=[]
        lid_l={}
        configurations[Role.STORAGED] = configuration

        platform._get_nodes(e_host)[host].set_configurations(configurations)

        for lconfig in configuration.listens:
            lid_l = OrderedDict([
                ('addr', lconfig.addr),
                ('port', lconfig.port)
            ])
            sid_l[host].append(lid_l)

        ordered_puts(sid_l)


def dispatch(args):
    p = Platform(args.exportd, Role.EXPORTD | Role.STORAGED)
    globals()[args.action.replace('-', '_')](p, args)
